src/aspire/source/micrograph.py

```diff
@@ -0,0 +1,209 @@
+                count += 1
+        # Check for zeroes
+        zero_count = 0
+        for center in centers:
+            if center[0] == -9999 and center[1] == -9999:
+                zero_count += 1
+        if zero_count > 0:
+            raise RuntimeError("Not enough centers generated.")
+        return centers
+
+    def _generate_center(self):
+        parity = (self.boundary + 1) % 2
+        x = (
+            (self.micrograph_size - 2 * self.boundary - parity) * np.random.rand()
+            + self.boundary
+            + parity
+        )
+        y = (
+            (self.micrograph_size - 2 * self.boundary - parity) * np.random.rand()
+            + self.boundary
+            + parity
+        )
+        return (int(x), int(y))
+
+    def _in_boundary(self, x, y):
+        return (
+            x - self.L // 2 > self.boundary
+            and x + self.L // 2 < self.micrograph_size - self.boundary
+            and y - self.L // 2 > self.boundary
+            and y + self.L // 2 < self.micrograph_size - self.boundary
+        )
+
+    def __len__(self):
+        """ """
+        return self.n
+
+    def clean_micrographs(self):
+        return self._clean_micrographs_accessor
+
+    def micrographs(self):
+        return self._micrographs_accessor
+
+    def _micrographs(self, indices):
+        micrograph = self._clean_micrographs(indices)
+        if self.noise_adder:
+            micrograph = self.noise_adder.forward(micrograph)
+        return micrograph
+
+    def _clean_micrographs(self, indices):
+        # Initialize empty micrograph
+        clean_micrograph = np.zeros((self.micrograph_size, self.micrograph_size))
+        pad = 0
+        if self.boundary < 0:
+            pad = self.L
+            clean_micrograph = np.pad(
+                clean_micrograph, pad, "constant", constant_values=(0)
+            )
+        # Get centers
+        centers = self.centers[indices][0]
+        parity = self.L % 2
+        for i in range(centers.shape[0]):
+            image = self.simulation.clean_images[self.ppm * indices + i].asnumpy()
+            x_left = centers[i][0] - self.L // 2 + pad
+            x_right = centers[i][0] + self.L // 2 + parity + pad
+            y_left = centers[i][1] - self.L // 2 + pad
+            y_right = centers[i][1] + self.L // 2 + parity + pad
+            clean_micrograph[x_left:x_right, y_left:y_right] = (
+                clean_micrograph[x_left:x_right, y_left:y_right] - image
+            )
+
+        if self.boundary < 0:
+            clean_micrograph = clean_micrograph[
+                pad : self.micrograph_size + parity + pad,
+                pad : self.micrograph_size + parity + pad,
+            ]
+        return Image(clean_micrograph)
+
+    def _images(self, indices):
+        return self.simulation.images[indices].asnumpy()[0]
+
+    def get_micrograph(self, id):
+        """
+        :param id: Global ID of the particle
+        """
+        if id >= self.ppm * self.n:
```

Comment:
I think you'll want to precumpute `ppm*n` and save it.  `total_particle_count`.

You'll use it elsewhere.